### PR TITLE
Consensus from bam

### PR DESCRIPTION
**Pull request by FelixMoelder**
This PR implements consensus read calling on BAM-files.
Consensus reads are calculated for all reads being duplicates of each other, as well as for all overlapping read mates.

As input a BAM-file with duplicates marked by PicardTools is required.
When marking duplicates the option `TAG_DUPLICATE_SET_MEMBERS` must be used to identify reads of the same duplicate group.

For each duplicate group consensus reads are called.
In case mate reads exists these will be included in the consensus if mates overlap.

The overlap is not identified by the insert size but by the CIGAR-String as soft clips need to be considered.

For each duplicate group reads will be additionally clustered using starcode which may result in multiple consensus per duplicate group.

Reads are be written to BAM format.
Consensus reads will be unmapped, unique reads stay unchanged.

## Todos:

- [x] Revise overlap calculation (Consider special cases)
- [x] Define template for unmapped consensus reads (Flags, Tags, CIGAR)
- [x] Cargo.toml: Replace rust-bio and rust-htslib by newest release (rust-bio needs new release of current master, rust-htslib needs release of [PR](https://github.com/rust-bio/rust-htslib/pull/136))

### CODE REVIEW

**Comment by johanneskoester** [comment]
Rust-Bio 0.27 is released.

**Comment by FelixMoelder** [comment]
> Rust-Bio 0.27 is released.

Perfect, thanks!
Than we just need a new release of rust-htslib. I guess that should be a major release as the `set()` function parameters have changed and this could potentially break existing rust code. ;)

**Comment by FelixMoelder** [comment]
All changes from rust-bio und rust-htslib are now implemented and (hopefully) everything is running.
Time to review. ;)

File: README.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ Rust-Bio-Tools provides a command `rbt`, which currently supports the following
 * a linear time extraction of depth information from BAMs at given loci (`rbt bam-depth`)
 * a utility to quickly filter records from a FASTQ file (`rbt fastq-filter`)
 * a tool to merge FASTQ reads using unique molecular identifiers (UMIs) (`rbt call-consensus-reads`)
+* a tool to merge BAM reads based on duplicates marked by [Picards Tools](https://broadinstitute.github.io/picard/) (`rbt call-consensus-from-bam`)
```
Comment (johanneskoester) [inline README.md]: Maybe rather two subcommands?
`rbt call-consensus-reads [bam|fastq] ...`

File: src/bam/call_consensus_reads/mod.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+use std::error::Error;
+
+pub fn call_consensus_reads_from_paths(
+    bam_in: &str,
+    bam_out: &str,
+    seq_dist: usize,
+    verbose_read_names: bool,
+) -> Result<(), Box<dyn Error>> {
+    eprintln!("Reading input files:\n    {}", bam_in);
+    eprintln!("Writing output to:\n    {}", bam_out);
```
Comment (johanneskoester) [inline src/bam/call_consensus_reads/mod.rs]: Let us rather use rust logging here: `info!(...)`.

File: src/common_functions/mod.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
+use bio::stats::probs::{LogProb, PHREDProb};
```
Comment (johanneskoester) [inline src/common_functions/mod.rs]: I'd rather just call this module `src/common.rs`.

File: src/common_functions/mod.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+pub trait CalcConsensus<'a, R: SequenceRead> {
+    fn validate_read_lengths(recs: &[R]) -> bool {
+        let reference_length = recs[0].len();
+        recs.iter()
+            .map(|rec| rec.len())
+            .all(|len| len == reference_length)
+    }
+
+    #[allow(unused_doc_comments)]
```
Comment (johanneskoester) [inline src/common_functions/mod.rs]: Are you sure this attribute is necessary? I think traits may have doc comments.

Comment (FelixMoelder) [reply src/common_functions/mod.rs]: You are right. That was left over from the previous separate implementation.

File: README.md (hunk 1 above)
Comment (FelixMoelder) [reply README.md]: Good point. That's changed now.

**Pull request merged**